allocate.go

```diff
@@ -0,0 +1,180 @@
+	peer "github.com/libp2p/go-libp2p-peer"
+
+	"github.com/ipfs/ipfs-cluster/api"
+)
+
+// This file gathers allocation logic used when pinning or re-pinning
+// to find which peers should be allocated to a Cid. Allocation is constrained
+// by ReplicationFactorMin and ReplicationFactorMax parametres obtained
+// from the Pin object.
+
+//The allocation
+// process has several steps:
+//
+// * Find which peers are pinning a CID
+// * Obtain the last values for the configured informer metrics from the
+//   monitor component
+// * Divide the metrics between "current" (peers already pinning the CID)
+//   and "candidates" (peers that could pin the CID), as long as their metrics
+//   are valid.
+// * Given the candidates:
+//   * Check if we are overpinning an item
+//   * Check if there are not enough candidates for the "needed" replication
+//     factor.
+//   * If there are enough candidates:
+//     * Call the configured allocator, which sorts the candidates (and
+//       may veto some depending on the allocation strategy.
+//     * The allocator returns a list of final candidate peers sorted by
+//       order of preference.
+//     * Take as many final candidates from the list as we can, until
+//       ReplicationFactorMax is reached. Error if there are less than
+//       ReplicationFactorMin.
+
+// allocate finds peers to allocate a hash using the informer and the monitor
+// it should only be used with valid replicationFactors (rplMin and rplMax
+// which are positive and rplMin < rplMax).
+func (c *Cluster) allocate(hash *cid.Cid, rplMin, rplMax int, blacklist []peer.ID) ([]peer.ID, error) {
+	// Figure out who is holding the CID
+	pinAllocations := c.getCurrentAllocations(hash)
+	metrics, err := c.getInformerMetrics()
+	if err != nil {
+		return nil, err
+	}
+
+	current := make(map[peer.ID]api.Metric)
+	candidates := make(map[peer.ID]api.Metric)
+
+	// Divide metrics between current and candidates.
+	for _, m := range metrics {
+		switch {
+		case m.Discard() || containsPeer(blacklist, m.Peer):
+			// blacklisted peers do not exist for us
+			continue
+		case containsPeer(pinAllocations, m.Peer):
+			current[m.Peer] = m
+		default:
+			candidates[m.Peer] = m
+		}
+	}
+
+	return c.obtainAllocations(hash,
+		rplMin,
+		rplMax,
+		current,
+		candidates)
+}
+
+// getCurrentAllocations returns the list of peers allocated to a Cid.
+func (c *Cluster) getCurrentAllocations(h *cid.Cid) []peer.ID {
+	var allocs []peer.ID
+	st, err := c.consensus.State()
+	if err != nil {
+		// no state we assume it is empty. If there was other
+		// problem, we would fail to commit anyway.
+		allocs = []peer.ID{}
+	} else {
+		pin := st.Get(h)
+		allocs = pin.Allocations
+	}
+	return allocs
+}
+
+// getInformerMetrics returns the MonitorLastMetrics() for the
+// configured informer.
+func (c *Cluster) getInformerMetrics() ([]api.Metric, error) {
+	var metrics []api.Metric
+	metricName := c.informer.Name()
+	l, err := c.consensus.Leader()
```

Comment:
Yeah, that is correct. Monitoring is the place where Leader() leaks out of of the consensus layer.

The other easy alternative is to broadcast all metrics everywhere from everywhere. Way more resilient, at the expense of possibly bad scaling. (although for the cluster sizes we work with it might be doable).

We should come up with a better way to broadcast things like metrics though. Perhaps a ring-fashion where every peer forwards the metrics it knows about to the next one in the list. This will come up as a problem at some point I'm sure.